nexus-writer/src/nexus/builder.rs

```diff
@@ -0,0 +1,237 @@
+use super::{Run, RunParameters};
+use crate::GenericEventMessage;
+use anyhow::{anyhow, Error, Result};
+use chrono::Duration;
+#[cfg(test)]
+use std::collections::vec_deque;
+use std::{collections::VecDeque, path::PathBuf};
+use supermusr_streaming_types::{
+    ecs_6s4t_run_stop_generated::RunStop, ecs_pl72_run_start_generated::RunStart,
+};
+use tracing::warn;
+
+#[derive(Default, Debug)]
+pub(crate) struct Nexus {
+    filename: Option<PathBuf>,
+    run_cache: VecDeque<Run>,
+    run_number: u32,
+}
+
+impl Nexus {
+    pub(crate) fn new(filename: Option<PathBuf>) -> Self {
+        Self {
+            filename,
+            ..Default::default()
+        }
+    }
+
+    fn append_context(&self, e: Error) -> Error {
+        anyhow!(
+            "\nNexus Context: {0:?}\n{e}",
+            self.run_cache
+                .iter()
+                .map(|run| run.get_name())
+                .collect::<Vec<_>>()
+        )
+    }
+
+    #[cfg(test)]
+    fn cache_iter(&self) -> vec_deque::Iter<'_, Run> {
```

Comment:
Why is this needed?

Reply:
This is used in tests to check runs have been successfully created.